File: CStructures/CStructures/CircularLinkedList.cpp
#include "pch.h"
#include "CircularLinkedList.h"
#include <iostream>


CircularLinkedList::CircularLinkedList()
{
	tail = NULL;
	listSize = 0;
}

int CircularLinkedList::size() { return listSize; }

bool CircularLinkedList::isEmpty() { return listSize == 0; }

int CircularLinkedList::first() {
	if (isEmpty()) {
		return NULL;
	}
	else {
		return tail->getNext()->getElement();
	}
}

int CircularLinkedList::last() {
	if (isEmpty()) {
		return NULL;
	}
	else {
		return tail->getElement();
	}
}

void CircularLinkedList::rotate() {
	if (tail != NULL) {
		tail = tail->getNext();
	}
}

void CircularLinkedList::addFirst(int n) {
	if (listSize == 0) {
		tail = new Node(n , NULL);
		tail->setNext(tail);
	}
	else {
		node newest = new Node(n, tail->getNext());
		tail->setNext(newest);
	}
	listSize++;
}

void CircularLinkedList::addLast(int n) {
	addFirst(n);
	tail = tail->getNext();
}

int CircularLinkedList::removeFirst() {
	if (isEmpty()) {
		return NULL;
	}
	else {
		Node *head = tail->getNext();
		if (tail == head) {
			tail = NULL;
		}
		else {
			tail->setNext(head->getNext());
		}
		listSize--;
		return head->getElement();
	}
}

bool CircularLinkedList::search(int reference) {
	node auxiliar = tail->getNext();	
	bool flag = true;
	while (auxiliar != tail && flag)
	{
		if (auxiliar->getElement() == reference)
		{
			flag = false;
		}
		else
		{
			auxiliar = auxiliar->getNext();
		}
	}

	if (auxiliar == tail && auxiliar->getElement() == reference)
	{
		return flag;
	}
	else
	{
		return !flag;
	}
}

void CircularLinkedList::showElements() {
	node auxiliar = tail->getNext();
	while (auxiliar != tail)
	{
		std::cout << auxiliar->getElement() << "\t";
		auxiliar = auxiliar->getNext();
	}
	std::cout << auxiliar->getElement() << "\n";
}

CircularLinkedList::~CircularLinkedList()
{
}

File: DynamicStructures/EstrcuturasDinamicas/CircularLinkedList.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace EstrcuturasDinamicas
{
    public class CircularLinkedList<T>
    {
        private class Node<T>
        {
            private T element;
            private Node<T> next;
            public Node(T element, Node<T> next)
            {                
                this.element = element;
                this.next = next;
            }
            public T getElement()
            {
                return element;
            }
            public Node<T> getNext()
            {
                return next;
            }
            public void setNext(Node<T> next)
            {
                this.next = next;
            }
        }


        private Node<T> tail = null;
        private int listSize = 0;

        public int size() { return listSize; }

        public bool isEmpty() { return listSize == 0; }

        public T first()
        {
            if (isEmpty()) return default(T);
            return tail.getNext().getElement();//Primer cambio
        }

        public T last()
        {
            if (isEmpty()) return default(T);
            return tail.getElement();
        }

        public void rotate()
        {
            if (tail != null)
                tail = tail.getNext();
        }

        public void addFirst(T e)
        {
            if (listSize == 0)
            {
                tail = new Node<T>(e, null);
                tail.setNext(tail);
            }
            else
            {
                Node<T> newest = new Node<T>(e, tail.getNext());
                tail.setNext(newest);
            }
            listSize++;
        }

        public void addLast(T e)
        {
            addFirst(e);
            tail = tail.getNext();
        }

        public T removeFirst()
        {
            if (isEmpty()) return default(T);
            Node<T> head = tail.getNext();
            if (head == tail) tail = null;
            else tail.setNext(head.getNext());
            listSize--;
            return head.getElement();
        }


    }
}
